### PR TITLE
feat: add model hook decorators

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

This PR adds the `@sequelize/core/decorators-legacy` import. Called `decorators-legacy` because these decorators are implemented using the legacy decorators proposal. Once TypeScript supports the new decorators proposal, we'll look into adding `@sequelize/core/decorators`

This PR adds the hook decorators from https://github.com/sequelize/sequelize-typescript/. More to come

### CODE REVIEW

File: src/decorators/legacy/model-hooks-single.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
+import type { HookOptions, HookDecoratorArgs } from './hook-decorators.js';
```
Comment (WikiRik) [inline src/decorators/legacy/model-hooks-single.ts:1]: Looking at https://github.com/sequelize/sequelize/blob/a4f8e62dd5b5b1c9df983b201581165d8465f107/src/model-typescript.ts#L102-L118 I'm still missing the `Associate` hooks

File: test/unit/decorators/hooks.test.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  BeforeRestore,
+  BeforeSave,
+  BeforeUpdate,
+  BeforeUpsert,
+  BeforeValidate,
+} from '@sequelize/core/decorators-legacy';
+import { expect } from 'chai';
+
+// map of hook name to hook decorator
+const hookMap: Partial<Record<keyof ModelHooks, Function>> = {
```
Comment (WikiRik) [inline test/unit/decorators/hooks.test.ts:35]: `ValidationFailed` and the two `Sync` ones are missing here (apart from the `Associate` ones not even implemented)

File: src/decorators/legacy/index.mjs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import Pkg from './index.js';
```
Comment (WikiRik) [inline src/decorators/legacy/index.mjs:1]: Don't forget to export the `Associate` ones as well

**Review by WikiRik — COMMENTED** [review]
What's the plan for merging the test suite of `sequelize-typescript`? Since there are two files dedicated to hooks but they also seem to use other decorators; https://github.com/sequelize/sequelize-typescript/blob/master/test/models/Hook.ts and https://github.com/sequelize/sequelize-typescript/blob/master/test/specs/hooks/hooks.spec.ts

File: src/decorators/legacy/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# decorators-legacy
+
+This directory regroups the decorators that are built using the legacy decorator proposal.
```
[the remark this answers is not in the document]
Comment (ephys) [reply src/decorators/legacy/README.md:3]: None

**Comment by ephys** [comment]
Associate is missing by accident but there are a bunch of decorators in sequelize-typescript that can't be used. For instance, `@BeforeConnect` can't be used on a model, the event is triggered on the Sequelize instance, so having a decorator for it isn't possible (or I'm missing something)

**Comment by WikiRik** [comment]
Yeah, I assumed so. I only looked at the `validModelHooks` list that we have

File: src/decorators/legacy/model-hooks-single.ts (hunk 1 above)
Comment (ephys) [reply src/decorators/legacy/model-hooks-single.ts:1]: Done

File: test/unit/decorators/hooks.test.ts (hunk 2 above)
Comment (ephys) [reply test/unit/decorators/hooks.test.ts:40]: Done

File: src/decorators/legacy/index.mjs (hunk 3 above)
Comment (ephys) [reply src/decorators/legacy/index.mjs:1]: Done

**Comment by ephys** [comment]
Added the missing tests & fixed what failed

**Review by WikiRik — APPROVED** [review]

**Pull request merged**